### PR TITLE
Reimplementing target description concept using DLTI Attributes

### PR DESCRIPTION
**Pull request by nhasabni**
and Interfaces. This is a newer implementation of PR #85141 and [RFC](https://discourse.llvm.org/t/rfc-target-description-and-cost-model-in-mlir/76990) by considering reviews and comments on the original PR.

As an example of attributes supported by this commit:

    ```
    module attributes {
       dlti.tsd_spec =
         #dlti.tdd_spec<
           #dlti.dl_entry<"dlti.device_id", 0: ui32>,
           #dlti.dl_entry<"dlti.device_type", "CPU">,
           #dlti.dl_entry<"dlti.canonicalizer_max_iterations", 100 : i32>,
           #dlti.dl_entry<"dlti.canonicalizer_max_num_rewrites", -5 : i32>>,
        #dlti.tdd_spec<
           #dlti.dl_entry<"dlti.device_id", 1: ui32>,
           #dlti.dl_entry<"dlti.device_type", "GPU">,
           #dlti.dl_entry<"dlti.max_vector_op_width", 64 : ui32>>,
        #dlti.tdd_spec<
           #dlti.dl_entry<"dlti.device_id", 2: ui32>,
           #dlti.dl_entry<"dlti.device_type", "XPU">>>
    }
    ```

### CODE REVIEW

**Comment by nhasabni** [comment]
@rengolin @joker-eph Please take a look. This is a DLTI based implementation for representing a target description as attributes.

File: mlir/include/mlir/Dialect/DLTI/DLTI.h [hunk 1]
```diff
@@ -124,6 +126,141 @@ class DataLayoutSpecAttr
   static constexpr StringLiteral name = "builtin.data_layout_spec";
 };
 
+//===----------------------------------------------------------------------===//
+// TargetSystemDescSpecAttr
+//===----------------------------------------------------------------------===//
+
+/// A system description attribute is a list of device descriptors, each
+/// having a uniq device ID
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/DLTI/DLTI.h:134]:
```suggestion
/// having a unique device ID
```

File: mlir/include/mlir/Dialect/DLTI/DLTI.h [hunk 2]
```diff
@@ -124,6 +126,141 @@ class DataLayoutSpecAttr
...
+class TargetSystemDescSpecAttr
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/DLTI/DLTI.h:135]: Why can't this attribute be defined with TableGen?

File: mlir/lib/Transforms/Canonicalizer.cpp [hunk 3]
```diff
@@ -59,6 +75,41 @@ struct Canonicalizer : public impl::CanonicalizerBase<Canonicalizer> {
...
+      ModuleOp moduleOp = op->getParentOfType<ModuleOp>();
+      if (std::optional<int64_t> v =
+              DataLayout(moduleOp).getCanonicalizerMaxNumRewrites(cpuID)) {
+        config.maxNumRewrites = *v;
+      }
+    }
+    LLVM_DEBUG(llvm::dbgs() << "[CostModel] Canonicalizer MaxIterations (new):"
+                            << config.maxIterations << "\n");
+    LLVM_DEBUG(llvm::dbgs() << "[CostModel] Canonicalizer MaxNumRewrites (new):"
+                            << config.maxNumRewrites << "\n");
```
Comment (joker-eph) [inline mlir/lib/Transforms/Canonicalizer.cpp:111]: These aren't properties of the target, it's not clear to me why we'd do this through this mechanism.

Comment (nhasabni) [reply mlir/lib/Transforms/Canonicalizer.cpp:111]: This was just for demo purpose. I wanted to demo the infra that I am building using some hardcoded constants - hence chose these. Exactly which properties of target devices to capture looks a separate topic to me. BTW, I've shown other example of a device property in this PR also.

File: mlir/include/mlir/Dialect/DLTI/DLTI.h (hunk 2 above)
Comment (nhasabni) [reply mlir/include/mlir/Dialect/DLTI/DLTI.h:135]: I was following the same mechanism as it is currently in DLTI.

**Pull request closed without merging**